### PR TITLE
--first-parent filtration added, --full-history fixed, hotkeys exposed, context menu in file history window unification

### LINKED ISSUE #3215: Follow renames in file history is broken
I've just build locally latest GitExtensions from code. Looks nice but I found that now file history doesn't follow renames.

Current release 2.48.05:
Settings:
image
File history (master branch is selected since otherwise it shows tons of not relevant commits):
image

Trunk from 6/10/2016:
Settings:
image
File history (master branch is selected to be on-par with other screenshot):
image

### PR DESCRIPTION
**Pull request by FallenGameR**
This change addresses issues we encountered while using GitExtensions on a huge repository backed up by Microsoft VSTS that can complete pull requests with merges. Merge is done correctly on the server (first parent is always the previous commit is master), but looking at commit history locally in GitExtensions it was hard to read what have happened. Since repository size is huge, lots of merges make history unreadable. This change allows to see the relevant high-level changes and hide unessential ones.

- #3216 - Doesn't seem to be a way to get only --first-parent history
- #3215 - Follow renames in file history is broken 
- Along the way several related areas are fixed - several hotkeys that were registered now are shown in menu. 
- Also context menu in file history window now has View and Navigation sections as browse form do since they use the same control for showing commits. 

# --first-parent filtration added

#3216
- Command with icon was added to FormBrowse, FormFileHistory, main menu and context menus
- Button is checked when filtration is applied and unchecked when not
- When ShowFirstParent and ShowCurrentBranchOnly are working DvcsGraph don't draw extra-long merge lines
- On gotocommit command additional branch filtration is applied to make hidden commits traversable 
- SelectInitialRevision makes sure it don't jump to unavailable commit when --first-parent filtration is applied

Here is how it looks:

Show current branch only, no --first-parent (old)
Show current branch only, no --first-parent

Show current branch only, --first-parent enabled (new)
Show current branch only, --first-parent enabled

# --full-history filtration fixed
#3215
It incorrectly used switch --simplify-by-decoration that actually truncated history, making opposite from user expects when --full-history is used. Fixed.

# Navigation and View submenus added
Navigation and View submenus added to FileHisotyForm context menu since all commands are applicable to ReviewGrid there and hotkeys for these commands actually worked there anyway already. Here is how it looks now:
Navigation and View submenus added

# Hotkeys changes
The following commands now what hotkeys are assigned to them in menus:
- ShowAllBranches
- ShowCurrentBranchOnly
- ShowFilteredBranches
- ShowRemoteBranches
- ToggleShowMergeCommits
- ToggleRevisionCardLayout
- ShowFirstParent
- RevisionFilter

Note that ToggleRevisionCardLayout that was around in codebase since 2011 was not exposed before. It is very useful when merge commit filtration is applied and card layout get unexpectedly changed.

Added new hotkeys with keystrokes that follows pattern of similar hotkeys and don't add collisions with existing hotkeys:
- ToggleShowMergeCommits = Control + Shift + M
- ShowFilteredBranches = Control + Shift + T
- ShowRemoteBranches = Control + Shift + R
- ShowFirstParent = Control + Shift + S

Here is how it looks:
Hotkeys changes

# Other changes
Name refactoring for members with unclear or outdated names.

### CODE REVIEW

**Comment by jbialobr** [comment]
Can you add a setting for `--simplify-by-decoration` under the Detailed settings tab? It should be checked by default.

File: GitUI/CommandsDialogs/FormBrowse.cs [hunk 1]
```diff
@@ -3083,7 +3083,17 @@ private void RevisionInfo_CommandClick(object sender, CommitInfo.CommandEventArg
...
-                RevisionGrid.SetSelectedRevision(new GitRevision(Module, e.Data));
+                var revision = new GitRevision(Module, e.Data);
+                var found = RevisionGrid.SetSelectedRevision(revision);
+
+                // When 'git log --first-parent' filtration is used, user can click on child commit
+                // that is not present in the shown git log. User still wants to see the child commit
+                // and to make it possible we add explicit branch filter and refresh.
+                if (AppSettings.ShowFirstParent && !found)
+                {
+                    _filterBranchHelper.SetBranchFilter(e.Data, refresh: true);
```
Comment (jbialobr) [inline GitUI/CommandsDialogs/FormBrowse.cs]: I would use revision.Guid insetad of e.Data.

**Comment by jbialobr** [comment]
I think that [--simplify-by-decoration in FollowParentRewriter](https://github.com/gitextensions/gitextensions/blob/b512d3b235b6f84f8bfcd3efd0f111e2d1c1c8ba/GitUI/UserControls/RevisionGridClasses/FollowParentRewriter.cs#L183) should be synced with FormHistory

File: GitUI/CommandsDialogs/FormBrowse.cs (hunk 1 above)
Comment (FallenGameR) [inline GitUI/CommandsDialogs/FormBrowse.cs]: I've used the argument that was in the existing code. Should I change it?

**Comment by FallenGameR** [comment]
@jbialobr, honestly I don't know why --simplify-by-decoration is even needed:

> Can you add a setting for --simplify-by-decoration under the Detailed settings tab? It should be checked by default.
> I think that --simplify-by-decoration in FollowParentRewriter should be synced with FormHistory

In the existing code it is used only with --full-history. But it is doing something oposite - it truncates history as shown in #3215. Could you explain use case for this argument? IMHO it is not needed and harmful.

**Comment by jbialobr** [comment]
> Could you explain use case for this argument? IMHO it is not needed and harmful.

I assume that someone added this option because he needed it. I don't want to make him not using it because someone else thinks that this option is not needed or harmful (let him decide by himself).
I think that `--simplify-by-decoration` fits better under `View -> Set advanced filter` than Detailed settings.

File: GitUI/CommandsDialogs/FormBrowse.cs (hunk 1 above)
Comment (jbialobr) [inline GitUI/CommandsDialogs/FormBrowse.cs]: Yes, please.
`_filterBranchHelper.SetBranchFilter(revision.Guid, refresh: true);`
 is more descriptive than
`_filterBranchHelper.SetBranchFilter(e.Data, refresh: true);`

**Comment by FallenGameR** [comment]
@jbialobr, I can add to settings, but IMO it should be off by default. The reason is - in the latest published release there is no such switch. And the person who added it can't say why it is needed either (see discussion of #3215.).

**Comment by jbialobr** [comment]
That is great. Can you add it to the `View -> Set advanced filter` form?

**Comment by FallenGameR** [comment]
@jbialobr addressed both issues. Here is how --simplify-by-decoration looks like:

--simplify-by-decoration preview

**Comment by jbialobr** [comment]
What about https://github.com/gitextensions/gitextensions/pull/3250#issuecomment-233299169?

**Comment by FallenGameR** [comment]
@jbialobr could you give more context on the use case? In my experiments I didn't hit that code path ever. When this code is used for and when it is invoked? 

I can split so that FullHistoryInFileHistory would add only --full-history and SimplifyByDecoration would additionally add --simplify-by-decoration.

**Comment by jbialobr** [comment]
>I can split so that FullHistoryInFileHistory would add only --full-history and SimplifyByDecoration would additionally add --simplify-by-decoration.

I think that it is enough. @qgppl Can you confirm it is a decent change for https://github.com/gitextensions/gitextensions/pull/3250#issuecomment-233299169?

**Comment by FallenGameR** [comment]
@jbialobr feedback addressed and merged latest changes from master since there was a conflict in designer files.

**Comment by jbialobr** [comment]
@FallenGameR Thank you.

**Comment by FallenGameR** [comment]
Thank you for prompt code review!

**Pull request merged**